docs/docs/protocol-specs/gas-and-fees/kernel-tracking.md

```diff
@@ -0,0 +1,531 @@
+  +Field[MAX_NEW_NOTE_HASHES_PER_TX] new_note_hashes
+  +Field[MAX_NEW_NULLIFIERS_PER_TX] new_nullifiers
+  +PublicDataUpdateRequest[MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX] public_data_update_requests
+}
+CombinedAccumulatedData --> Gas
+
+class PublicContextInputs {
+    +Header historical_header
+    +GlobalVariables public_global_variables
+    +Gas gas_left
+    +Field transaction_fee
+}
+PublicContextInputs --> Header
+PublicContextInputs --> GlobalVariables
+PublicContextInputs --> Gas
+
+class GlobalVariables {
+    +GasFees gas_fees
+}
+GlobalVariables --> GasFees
+
+class GasFees {
+    +Fr fee_per_da_gas
+    +Fr fee_per_l2_gas
+}
+
+
+
+
+```
+
+## Public Context Initialization
+
+Whenever a public function is run, it has a `PublicContext` associated with it, which is initialized in part from a `PublicContextInputs` object.
+
+The sequencer must provide information including the `gas_left` and the `transaction_fee`, but we cannot trust these values to be correct: we must compute the correct values in the public kernel circuits, and validate that the sequencer provided the correct values.
+
+Further, the sequencer is only obligated to provide the `transaction_fee` to the teardown function, as that is the only point at which the transaction fee can be known.
+
+## Public Circuit Public Inputs
+
+The "outputs" of the public functions are coming from the public VM.
+
+Therefore, once we verify that the `start_gas_left` which the sequencer provided is correct, we can trust the `end_gas_left` that the public VM reports.
+
+Further, we can trust that the `transaction_fee` the public VM reported is the one which was made available to the public functions during teardown (though we must verify that the sequencer provided the correct value).
+
+## Public Kernel Setup
+
+The PublicKernelSetup circuit takes in a `PublicKernelData` and a `PublicCallData` and outputs a `PublicKernelCircuitPublicInputs`.
+
+It must assert that the `revert_code` in the `PublicCircuitPublicInputs` is equal to zero.
+
+It must compute the gas used in the `PublicKernelData` provided, and verify that the `gas_limits` in the  `PublicKernelData`'s `TxContext` *minus* the computed `gas_used` is equal to the `start_gas_left` specified on the `PublicCircuitPublicInputs`.
+
+This ensures that the public VM was provided with the correct starting gas values.
+
+It must update the gas used in `end_non_revertible` as:
+
+```rust
+# assume the previous PublicKernelCircuitPublicInputs was copied to circuit_outputs
+pub fn update_non_revertible_gas_used(public_call: PublicCallData, circuit_outputs: &mut PublicKernelCircuitPublicInputsBuilder) {
+    let tx_gas_limits = circuit_outputs.constants.tx_context.gas_settings.gas_limits;
+    let call_gas_left = public_call.call_stack_item.public_inputs.end_gas_left;
+    let accum_end_gas_used = circuit_outputs.end.gas_used;
+
+    circuit_outputs.end_non_revertible.gas_used = tx_gas_limits
```

Comment:
So this is saying, the `gas_used` **after** this function call is equal to `gas_limit - (gas_used_at_start_of_call + gas_used_during_call)`?

Reply:
Not quite. `call_gas_left` is equal to `gas_limit - (gas_used_at_start_of_call + gas_used_during_call)`

So this is saying the (non_revertible) `gas_used` here is equal to

```
gas_limit - call_gas_left - revertible_gas_used
= (gas_used_at_start_of_call + gas_used_during_call) - revertible_gas_used
```